src/graph.rs

```diff
@@ -130,6 +167,843 @@ impl Handle<Symbol> {
+        self.files.iter_handles()
+    }
+}
+
+impl Display for File {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.name)
+    }
+}
+
+impl Index<Handle<File>> for StackGraph {
+    type Output = File;
+    #[inline(always)]
+    fn index(&self, handle: Handle<File>) -> &File {
+        &self.files.get(handle)
+    }
+}
+
+#[doc(hidden)]
+pub struct DisplayFile<'a> {
+    wrapped: Handle<File>,
+    graph: &'a StackGraph,
+}
+
+impl<'a> Display for DisplayFile<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.graph[self.wrapped])
+    }
+}
+
+impl Handle<File> {
+    pub fn display(self, graph: &StackGraph) -> impl Display + '_ {
+        DisplayFile {
+            wrapped: self,
+            graph,
+        }
+    }
+}
+
+//-------------------------------------------------------------------------------------------------
+// Nodes
+
+/// Uniquely identifies a node in a stack graph.
+///
+/// Each node (except for the _root node_ and _jump to scope_ node) lives in a file, and has a
+/// _local ID_ that must be unique within its file.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct NodeID {
+    /// The file that the node comes from.
+    pub file: Handle<File>,
+    /// The unique identity of the node within its file.
+    pub local_id: u32,
+}
+
+#[doc(hidden)]
+pub struct DisplayNodeID<'a> {
+    wrapped: NodeID,
+    graph: &'a StackGraph,
+}
+
+impl<'a> Display for DisplayNodeID<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}({})",
+            self.wrapped.file.display(self.graph),
+            self.wrapped.local_id,
+        )
+    }
+}
+
+impl NodeID {
+    pub fn display(self, graph: &StackGraph) -> impl Display + '_ {
+        DisplayNodeID {
+            wrapped: self,
+            graph,
+        }
+    }
+}
+
+/// A node in a stack graph.
+pub enum Node {
+    DropScopes(DropScopesNode),
+    ExportedScope(ExportedScopeNode),
+    InternalScope(InternalScopeNode),
+    JumpTo(JumpToNode),
+    PushScopedSymbol(PushScopedSymbolNode),
+    PushSymbol(PushSymbolNode),
+    PopScopedSymbol(PopScopedSymbolNode),
+    PopSymbol(PopSymbolNode),
+    Root(RootNode),
+    Unknown(UnknownNode),
+}
+
+impl Node {
+    #[inline(always)]
+    pub fn is_definition(&self) -> bool {
+        match self {
+            Node::PopScopedSymbol(node) => node.is_definition,
+            Node::PopSymbol(node) => node.is_definition,
+            _ => false,
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_reference(&self) -> bool {
+        match self {
+            Node::PushScopedSymbol(node) => node.is_reference,
+            Node::PushSymbol(node) => node.is_reference,
+            _ => false,
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_jump_to(&self) -> bool {
+        matches!(self, Node::JumpTo(_))
+    }
```

Comment:
And they say Rust doesn't have reflection! Using this combination of `matches!` and `&self` is a neat trick.

Reply:
It's all compile-time!  No runtime reflection going on, this is just sugar for a `match` statement that returns `true` or `false`.